File: src/bgbg/engine/render/builder.py
"""Build a DisplayList for one frame. Stdlib only — this is the render logic.

Everything the canvas draws is decided here: the checkerboard, the image, the
faint object tints, the dim-outside, the glows, the marching ants, the layered
whole-vs-part highlight, the outline morph, the press ripple, the scan shimmer
and the cutout preview. A backend then just *replays* the list.

This file is the thing that must be mirrored in TypeScript, and the golden
corpus in `spec/` is what proves the mirror is faithful.

Each Push carries a single effect so the nesting order is explicit — a glow is a
blur *outside* a mask *outside* an opacity, and combining them into one node
would make that ambiguous.
"""
from ..geometry import ease_out
from .ops import (
    DisplayList, Push, FillRect, Checker, DrawImage, LinearGradient, StrokePath,
    Path, MaskSpec, translate, scale, rotate, scale_about,
)
from .scene import CELL, CHECK_LIGHT, CHECK_DARK, DIM, POINT_COLOR


# ---------- small composites (the vocabulary the original renderer used) -----
def _tint(mask, color, alpha, rect):
    """Paint `color` at `alpha` wherever the mask covers."""
    return Push(mask=MaskSpec([mask], rect, "keep"), children=[
        Push(opacity=alpha, children=[FillRect(rect, color)])])


def _glow(mask, color, alpha, gsk_radius, rect):
    """A soft halo: a blurred, masked colour fill.

    GSK's push_blur takes a radius; a Gaussian sigma is half of it. We carry the
    sigma because that is what CSS/Canvas2D speak, and the GSK backend doubles it
    back — so the two backends agree by construction rather than by luck.
    """
    return Push(blur_sigma=gsk_radius / 2.0,
                children=[_tint(mask, color, alpha, rect)])


def _dim_outside(masks, rect, alpha):
    """Darken everything the masks do NOT cover."""
    return Push(mask=MaskSpec(masks, rect, "cut"), children=[
        Push(opacity=alpha, children=[FillRect(rect, DIM)])])


def _ants(polys, color, view_scale, ant_phase):
    """Marching ants. Width and dash are divided by the view scale so they stay
    screen-constant as you zoom."""
    if not polys:
        return None
    return StrokePath(
        Path(polys=polys), color,
        width=max(0.6, 1.6 / view_scale),
        dash=(6.0 / view_scale, 4.5 / view_scale),
        dash_offset=-ant_phase / view_scale)


def _with_alpha(color, a):
    return (color[0], color[1], color[2], color[3] * max(0.0, min(1.0, a)))


def _blend(a, b, t):
    return tuple(a[i] + (b[i] - a[i]) * t for i in range(4))


# ---------- the pieces -------------------------------------------------------
def _shimmer(scan_phase, rect, iw, ih):
    """A soft diagonal highlight sweeping over the image — the "scanning" cue.
    p is offset so the band enters and leaves rather than wrapping abruptly."""
    p = -0.20 + 1.40 * scan_phase
    band = 0.13

    def stop(off, a):
        return (min(1.0, max(0.0, off)), (1.0, 1.0, 1.0, a))

    stops = [stop(0.0, 0.0), stop(p - band, 0.0), stop(p, 0.18),
             stop(p + band, 0.0), stop(1.0, 0.0)]
    return Push(opacity=0.9, children=[
        LinearGradient(rect, (0, 0), (iw, ih), stops)])


def _composite(sc, rect):
    """Result-panel cutout: the source clipped to the union of selected masks."""
    if not sc.clip_masks:
        return []
    out = []
    if sc.clip_bg is not None:
        out.append(Push(mask=MaskSpec(sc.clip_masks, rect, "keep"),
                        children=[FillRect(rect, sc.clip_bg)]))
    out.append(Push(mask=MaskSpec(sc.clip_masks, rect, "keep"),
                    children=[DrawImage(sc.image, rect)]))
    return out


def _press(sc, anim, now, rect, view_scale):
    """The press "swizzle": press-scale, an intensified glow that decays after
    release, and expanding waves clipped to the object."""
    oid = anim.press_obj
    mask = sc.masks.get(oid)
    p = anim.press(now)
    if mask is None or p is None or not anim.press_pt:
        return []
    col = sc.colors.get(oid, POINT_COLOR)
    iw, ih = sc.image_size
    cx, cy = sc.centroids.get(oid) or (iw / 2.0, ih / 2.0)

    inner = [_glow(mask, col, p.glow, 26.0, rect),
             _tint(mask, col, p.tint, rect)]
    ant_col = col if (p.held or p.fade >= 1.0) else _with_alpha(col, p.fade)
    a = _ants(sc.polys.get(oid), ant_col, view_scale, anim.ant)
    if a is not None:
        inner.append(a)

    # rings, clipped to the object's silhouette
    px, py = anim.press_pt
    maxr = sc.radius.get(oid, 160.0)
    rings = []
    for ph, alpha in anim.waves(now):
        r = 8.0 + ease_out(ph) * maxr
        rings.append(StrokePath(Path(circles=[((px, py), r)]),
                                (1.0, 1.0, 1.0, alpha),
                                width=max(1.0, 3.0 / view_scale)))
    if rings:
        inner.append(Push(mask=MaskSpec([mask], rect, "keep"), children=rings))

    xform = scale_about(p.scale, p.scale, cx, cy) if abs(p.scale - 1.0) > 1e-4 else ()
    return [Push(transform=xform, children=inner)]


def _morph_outline(sc, e, view_scale, ant_phase):
    """The focused object's outline mid-tween: lerp the ring on a switch, bloom
    from the centroid on enter, collapse to it on leave; crossfade the rest."""
    m = sc.morph
    if m is None:
        return []
    a, b = m.from_ring, m.to_ring
    if a is not None and b is not None:                 # switch
        ring = [(a[i][0] + (b[i][0] - a[i][0]) * e,
                 a[i][1] + (b[i][1] - a[i][1]) * e) for i in range(len(a))]
        col, env = _blend(m.cf, m.ct, e), 1.0
    elif b is not None:                                 # enter: bloom
        cx, cy = m.cen_to or b[0]
        ring = [(cx + (p[0] - cx) * e, cy + (p[1] - cy) * e) for p in b]
        col, env = m.ct, e
    elif a is not None:                                 # leave: collapse
        cx, cy = m.cen_from or a[0]
        ring = [(p[0] + (cx - p[0]) * e, p[1] + (cy - p[1]) * e) for p in a]
        col, env = m.cf, 1.0 - e
    else:
        return []

    out = []
    s = _ants([ring], _with_alpha(col, env), view_scale, ant_phase)
    if s is not None:
        out.append(s)
    if m.sec_from:
        s = _ants(m.sec_from, _with_alpha(m.cf, 1.0 - e), view_scale, ant_phase)
        if s is not None:
            out.append(s)
    if m.sec_to:
        s = _ants(m.sec_to, _with_alpha(m.ct, e), view_scale, ant_phase)
        if s is not None:
            out.append(s)
    return out


def _focus_outline(sc, anim, now, view_scale, hover, sel):
    """The focused outline: morphing while the focus changes, else static."""
    e = anim.morph_progress(now)
    if e is not None:
        return _morph_outline(sc, e, view_scale, anim.ant)
    if hover and hover in sc.masks and hover not in sel:
        col = sc.colors.get(hover, POINT_COLOR)
        s = _ants(sc.polys.get(hover), col, view_scale, anim.ant)
        if s is not None:
            return [s]
    return []


def _seg(sc, anim, now, rect, view_scale):
    """The segmentation overlay, in draw order."""
    sel, hover = sc.selected, sc.hover_id
    gen, spec = sc.hover_gen, sc.hover_spec
    # Paint the selection in id order. `sel` is a set, and a set's iteration
    # order is its hash order in Python but its insertion order in JS — so
    # relying on it would make the two cores emit ops in different orders for
    # the same scene. `sel` stays the set, for O(1) membership.
    sel_order = sorted(sel)
    # Over a stack, light up the whole (general) and the part (specific) as
    # distinct colour layers rather than one flat highlight.
    layered = (sc.hover_depth >= 2 and gen and spec and gen != spec
               and gen in sc.masks and spec in sc.masks
               and gen not in sel and spec not in sel)
    pulse = 0.35 + 0.65 * anim.pulse
    skip = {gen, spec} if layered else {hover}
    out = []

    # 1. faint tint on every non-selected, non-focused object
    for oid, mask in sc.masks.items():
        if oid in sel or oid in skip:
            continue
        col = sc.colors.get(oid)
        if col is not None:
            out.append(_tint(mask, col, 0.12, rect))

    # 2. dim everything outside the selection
    if sc.masks and sel:
        out.append(_dim_outside([sc.masks[o] for o in sel_order], rect, 0.55))

    # 3. selected objects — glow + fill + ants, with a tactile "pop"
    iw, ih = sc.image_size
    for oid in sel_order:
        mask, col = sc.masks.get(oid), sc.colors.get(oid)
        if mask is None or col is None:
            continue
        pop = anim.pop_scale(oid, now)
        kids = [_glow(mask, col, 0.22 + 0.20 * anim.pulse, 22.0, rect),
                _tint(mask, col, 0.42, rect)]
        a = _ants(sc.polys.get(oid), col, view_scale, anim.ant)
        if a is not None:
            kids.append(a)
        if pop != 1.0:
            cx, cy = sc.centroids.get(oid) or (iw / 2.0, ih / 2.0)
            out.append(Push(transform=scale_about(pop, pop, cx, cy), children=kids))
        else:
            out.extend(kids)

    # 4. hovered (not selected): the whole-vs-part layers, or a single highlight
    if layered:
        gmask, gcol = sc.masks[gen], sc.colors[gen]
        smask, scol = sc.masks[spec], sc.colors[spec]
        out.append(_dim_outside([gmask], rect, 0.5))
        out.append(_glow(gmask, gcol, 0.14 * pulse, 22.0, rect))
        out.append(_tint(gmask, gcol, 0.18, rect))       # whole, beneath
        out.append(_glow(smask, scol, 0.26 * pulse, 18.0, rect))
        out.append(_tint(smask, scol, 0.40, rect))       # part, on top
    elif hover and hover in sc.masks and hover not in sel:
        col = sc.colors.get(hover, POINT_COLOR)
        out.append(_glow(sc.masks[hover], col, 0.28 * pulse, 20.0, rect))
        out.append(_tint(sc.masks[hover], col, 0.34, rect))

    out.extend(_focus_outline(sc, anim, now, view_scale, hover, sel))

    # 5. point mode — dim outside the object, glow + tint + ants
    if sc.point_mask is not None:
        col = POINT_COLOR
        out.append(_dim_outside([sc.point_mask], rect, 0.55))
        out.append(_glow(sc.point_mask, col, 0.20 + 0.18 * anim.pulse, 20.0, rect))
        out.append(_tint(sc.point_mask, col, 0.30, rect))
        a = _ants(sc.point_polys, col, view_scale, anim.ant)
        if a is not None:
            out.append(a)

    # 6. press-and-hold ripple
    if anim.press_obj in sc.masks and anim.press_pt:
        out.extend(_press(sc, anim, now, rect, view_scale))

    return out


# ---------- the frame --------------------------------------------------------
def build(sc, pane, anim, now):
    """The whole frame, as a DisplayList."""
    vw, vh = pane.view_w, pane.view_h
    ops = [Checker((0, 0, vw, vh), CELL, CHECK_LIGHT, CHECK_DARK)]

    ew, eh = pane.effective_size()
    if sc.image is None or not ew or not eh or vw <= 0 or vh <= 0:
        return DisplayList(ops, scale=1.0, view_size=(vw, vh))

    s = pane.scale()
    if not s:
        return DisplayList(ops, scale=1.0, view_size=(vw, vh))

    iw, ih = sc.image_size
    rect = (0, 0, iw, ih)

    xform = [translate(vw / 2 + pane.ox, vh / 2 + pane.oy), scale(s, s)]
    if pane.rot:
        xform.append(rotate(pane.rot * 90))
    xform.append(scale(-1 if pane.fh else 1, -1 if pane.fv else 1))
    xform.append(translate(-iw / 2, -ih / 2))

    inner = []
    if sc.clip_active:
        inner.extend(_composite(sc, rect))
    else:
        inner.append(DrawImage(sc.image, rect, "smooth"))
        if anim.scanning:
            inner.append(_shimmer(anim.scan_phase, rect, iw, ih))
        if sc.seg_mode:
            seg = _seg(sc, anim, now, rect, s)
            if anim.reveal < 1.0:            # fade + gentle scale-in of overlays
                z = 0.97 + 0.03 * anim.reveal
                inner.append(Push(opacity=max(0.0, anim.reveal), children=[
                    Push(transform=scale_about(z, z, iw / 2, ih / 2),
                         children=seg)]))
            else:
                inner.extend(seg)

    ops.append(Push(transform=xform, children=inner))
    return DisplayList(ops, scale=s, view_size=(vw, vh),
                       animating=anim.needs_tick())

File: web/core/render/builder.ts
/**
 * Build a DisplayList for one frame — the mirror of engine/render/builder.py.
 *
 * This is THE file that has to be faithful: everything the canvas draws is
 * decided here, and spec/goldens/display_list/*.json is the proof. Op order
 * matters as much as op content, which is why the selection is painted in sorted
 * id order (a Python set iterates by hash, a JS Set by insertion — relying on
 * either would make the two cores disagree for the same scene).
 */
import { easeOut, type Poly, type Vec2 } from "../geometry";
import type { RGBA } from "../color";
import { AnimState } from "../anim";
import { Pane } from "../pane";
import {
  DisplayList,
  Push,
  FillRect,
  Checker,
  DrawImage,
  LinearGradient,
  StrokePath,
  Path,
  MaskSpec,
  translate,
  scale,
  rotate,
  scaleAbout,
  type Handle,
  type Op,
  type Rect,
  type Step,
} from "./ops";
import {
  Scene,
  CELL,
  CHECK_LIGHT,
  CHECK_DARK,
  DIM,
  POINT_COLOR,
} from "./scene";

// ---------- small composites ----------
function tint(mask: Handle, color: RGBA, alpha: number, rect: Rect): Push {
  return new Push([new Push([new FillRect(rect, color)], [], alpha)], [], 1.0, 0.0,
    new MaskSpec([mask], rect, "keep"));
}

function glow(
  mask: Handle,
  color: RGBA,
  alpha: number,
  gskRadius: number,
  rect: Rect,
): Push {
  // GSK's push_blur takes a radius; a Gaussian sigma is half of it. We carry the
  // sigma because that is what CSS/Canvas2D speak.
  return new Push([tint(mask, color, alpha, rect)], [], 1.0, gskRadius / 2.0);
}

function dimOutside(masks: Handle[], rect: Rect, alpha: number): Push {
  return new Push([new Push([new FillRect(rect, DIM)], [], alpha)], [], 1.0, 0.0,
    new MaskSpec(masks, rect, "cut"));
}

/** Marching ants: width and dash divided by the view scale to stay screen-constant. */
function ants(
  polys: Poly[] | undefined,
  color: RGBA,
  viewScale: number,
  antPhase: number,
): StrokePath | null {
  if (!polys || polys.length === 0) return null;
  return new StrokePath(
    new Path(polys),
    color,
    Math.max(0.6, 1.6 / viewScale),
    [6.0 / viewScale, 4.5 / viewScale],
    -antPhase / viewScale,
  );
}

function withAlpha(color: RGBA, a: number): RGBA {
  return [color[0], color[1], color[2], color[3] * Math.max(0.0, Math.min(1.0, a))];
}

function blend(a: RGBA, b: RGBA, t: number): RGBA {
  return [0, 1, 2, 3].map((i) => a[i] + (b[i] - a[i]) * t) as RGBA;
}

// ---------- the pieces ----------
function shimmer(scanPhase: number, rect: Rect, iw: number, ih: number): Push {
  const p = -0.2 + 1.4 * scanPhase;
  const band = 0.13;
  const stop = (off: number, a: number): [number, RGBA] => [
    Math.min(1.0, Math.max(0.0, off)),
    [1.0, 1.0, 1.0, a],
  ];
  const stops: [number, RGBA][] = [
    stop(0.0, 0.0),
    stop(p - band, 0.0),
    stop(p, 0.18),
    stop(p + band, 0.0),
    stop(1.0, 0.0),
  ];
  return new Push(
    [new LinearGradient(rect, [0, 0], [iw, ih], stops)],
    [],
    0.9,
  );
}

function composite(sc: Scene, rect: Rect): Op[] {
  if (!sc.clipMasks.length) return [];
  const out: Op[] = [];
  // The background fills the WHOLE rect and the cutout goes on top — it is a
  // background, not a backing for the subject.
  if (sc.clipBg !== null) out.push(new FillRect(rect, sc.clipBg));
  out.push(
    new Push([new DrawImage(sc.image!, rect)], [], 1.0, 0.0,
      new MaskSpec(sc.clipMasks, rect, "keep")),
  );
  return out;
}

function press(
  sc: Scene,
  anim: AnimState,
  now: number,
  rect: Rect,
  viewScale: number,
): Op[] {
  const oid = anim.pressObj;
  const mask = sc.masks.get(oid);
  const p = anim.press(now);
  if (mask === undefined || p === null || !anim.pressPt) return [];

  const col = sc.colors.get(oid) ?? POINT_COLOR;
  const [iw, ih] = sc.imageSize;
  const [cx, cy] = sc.centroids.get(oid) ?? [iw / 2.0, ih / 2.0];

  const inner: Op[] = [
    glow(mask, col, p.glow, 26.0, rect),
    tint(mask, col, p.tint, rect),
  ];
  const antCol = p.held || p.fade >= 1.0 ? col : withAlpha(col, p.fade);
  const a = ants(sc.polys.get(oid), antCol, viewScale, anim.ant);
  if (a !== null) inner.push(a);

  const [px, py] = anim.pressPt;
  const maxr = sc.radius.get(oid) ?? 160.0;
  const rings: Op[] = [];
  for (const [ph, alpha] of anim.waves(now)) {
    const r = 8.0 + easeOut(ph) * maxr;
    rings.push(
      new StrokePath(
        new Path([], [[[px, py], r]]),
        [1.0, 1.0, 1.0, alpha],
        Math.max(1.0, 3.0 / viewScale),
      ),
    );
  }
  if (rings.length) {
    inner.push(new Push(rings, [], 1.0, 0.0, new MaskSpec([mask], rect, "keep")));
  }

  const xform: Step[] =
    Math.abs(p.scale - 1.0) > 1e-4 ? scaleAbout(p.scale, p.scale, cx, cy) : [];
  return [new Push(inner, xform)];
}

function morphOutline(
  sc: Scene,
  e: number,
  viewScale: number,
  antPhase: number,
): Op[] {
  const m = sc.morph;
  if (m === null) return [];
  const a = m.fromRing;
  const b = m.toRing;
  let ring: Poly;
  let col: RGBA;
  let env: number;

  if (a !== null && b !== null) {
    // switch
    ring = a.map(
      (_, i) =>
        [
          a[i][0] + (b[i][0] - a[i][0]) * e,
          a[i][1] + (b[i][1] - a[i][1]) * e,
        ] as Vec2,
    );
    col = blend(m.cf!, m.ct!, e);
    env = 1.0;
  } else if (b !== null) {
    // enter: bloom
    const [cx, cy] = m.cenTo ?? b[0];
    ring = b.map((p) => [cx + (p[0] - cx) * e, cy + (p[1] - cy) * e] as Vec2);
    col = m.ct!;
    env = e;
  } else if (a !== null) {
    // leave: collapse
    const [cx, cy] = m.cenFrom ?? a[0];
    ring = a.map((p) => [p[0] + (cx - p[0]) * e, p[1] + (cy - p[1]) * e] as Vec2);
    col = m.cf!;
    env = 1.0 - e;
  } else {
    return [];
  }

  const out: Op[] = [];
  let s = ants([ring], withAlpha(col, env), viewScale, antPhase);
  if (s !== null) out.push(s);
  if (m.secFrom && m.secFrom.length) {
    s = ants(m.secFrom, withAlpha(m.cf!, 1.0 - e), viewScale, antPhase);
    if (s !== null) out.push(s);
  }
  if (m.secTo && m.secTo.length) {
    s = ants(m.secTo, withAlpha(m.ct!, e), viewScale, antPhase);
    if (s !== null) out.push(s);
  }
  return out;
}

function focusOutline(
  sc: Scene,
  anim: AnimState,
  now: number,
  viewScale: number,
  hover: number,
  sel: Set<number>,
): Op[] {
  const e = anim.morphProgress(now);
  if (e !== null) return morphOutline(sc, e, viewScale, anim.ant);
  if (hover && sc.masks.has(hover) && !sel.has(hover)) {
    const col = sc.colors.get(hover) ?? POINT_COLOR;
    const s = ants(sc.polys.get(hover), col, viewScale, anim.ant);
    if (s !== null) return [s];
  }
  return [];
}

function seg(
  sc: Scene,
  anim: AnimState,
  now: number,
  rect: Rect,
  viewScale: number,
): Op[] {
  const sel = sc.selected;
  const hover = sc.hoverId;
  const gen = sc.hoverGen;
  const spec = sc.hoverSpec;

  // Paint the selection in id order: a Python set iterates in hash order and a
  // JS Set in insertion order, so relying on either would make the two cores
  // emit ops in a different order for the same scene.
  const selOrder = [...sel].sort((a, b) => a - b);

  const layered =
    sc.hoverDepth >= 2 &&
    !!gen &&
    !!spec &&
    gen !== spec &&
    sc.masks.has(gen) &&
    sc.masks.has(spec) &&
    !sel.has(gen) &&
    !sel.has(spec);

  const pulse = 0.35 + 0.65 * anim.pulse;
  const skip = layered ? new Set([gen, spec]) : new Set([hover]);
  const out: Op[] = [];

  // 1. faint tint on every non-selected, non-focused object
  for (const [oid, mask] of sc.masks) {
    if (sel.has(oid) || skip.has(oid)) continue;
    const col = sc.colors.get(oid);
    if (col !== undefined) out.push(tint(mask, col, 0.12, rect));
  }

  // 2. dim everything outside the selection
  if (sc.masks.size && sel.size) {
    out.push(dimOutside(selOrder.map((o) => sc.masks.get(o)!), rect, 0.55));
  }

  // 3. selected objects — glow + fill + ants, with a tactile "pop"
  const [iw, ih] = sc.imageSize;
  for (const oid of selOrder) {
    const mask = sc.masks.get(oid);
    const col = sc.colors.get(oid);
    if (mask === undefined || col === undefined) continue;
    const pop = anim.popScale(oid, now);
    const kids: Op[] = [
      glow(mask, col, 0.22 + 0.2 * anim.pulse, 22.0, rect),
      tint(mask, col, 0.42, rect),
    ];
    const a = ants(sc.polys.get(oid), col, viewScale, anim.ant);
    if (a !== null) kids.push(a);
    if (pop !== 1.0) {
      const [cx, cy] = sc.centroids.get(oid) ?? [iw / 2.0, ih / 2.0];
      out.push(new Push(kids, scaleAbout(pop, pop, cx, cy)));
    } else {
      out.push(...kids);
    }
  }

  // 4. hovered (not selected): the whole-vs-part layers, or a single highlight
  if (layered) {
    const gmask = sc.masks.get(gen)!;
    const gcol = sc.colors.get(gen)!;
    const smask = sc.masks.get(spec)!;
    const scol = sc.colors.get(spec)!;
    out.push(dimOutside([gmask], rect, 0.5));
    out.push(glow(gmask, gcol, 0.14 * pulse, 22.0, rect));
    out.push(tint(gmask, gcol, 0.18, rect)); // whole, beneath
    out.push(glow(smask, scol, 0.26 * pulse, 18.0, rect));
    out.push(tint(smask, scol, 0.4, rect)); // part, on top
  } else if (hover && sc.masks.has(hover) && !sel.has(hover)) {
    const col = sc.colors.get(hover) ?? POINT_COLOR;
    out.push(glow(sc.masks.get(hover)!, col, 0.28 * pulse, 20.0, rect));
    out.push(tint(sc.masks.get(hover)!, col, 0.34, rect));
  }

  out.push(...focusOutline(sc, anim, now, viewScale, hover, sel));

  // 5. point mode
  if (sc.pointMask !== null) {
    const col = POINT_COLOR;
    out.push(dimOutside([sc.pointMask], rect, 0.55));
    out.push(glow(sc.pointMask, col, 0.2 + 0.18 * anim.pulse, 20.0, rect));
    out.push(tint(sc.pointMask, col, 0.3, rect));
    const a = ants(sc.pointPolys, col, viewScale, anim.ant);
    if (a !== null) out.push(a);
  }

  // 6. press-and-hold ripple
  if (sc.masks.has(anim.pressObj) && anim.pressPt) {
    out.push(...press(sc, anim, now, rect, viewScale));
  }

  return out;
}

// ---------- the frame ----------
export function build(
  sc: Scene,
  pane: Pane,
  anim: AnimState,
  now: number,
): DisplayList {
  const vw = pane.viewW;
  const vh = pane.viewH;
  const ops: Op[] = [
    new Checker([0, 0, vw, vh], CELL, CHECK_LIGHT, CHECK_DARK),
  ];

  const [ew, eh] = pane.effectiveSize();
  if (sc.image === null || !ew || !eh || vw <= 0 || vh <= 0) {
    return new DisplayList(ops, 1.0, [vw, vh]);
  }

  const s = pane.scale();
  if (!s) return new DisplayList(ops, 1.0, [vw, vh]);

  const [iw, ih] = sc.imageSize;
  const rect: Rect = [0, 0, iw, ih];

  const xform: Step[] = [
    translate(vw / 2 + pane.ox, vh / 2 + pane.oy),
    scale(s, s),
  ];
  if (pane.rot) xform.push(rotate(pane.rot * 90));
  xform.push(scale(pane.fh ? -1 : 1, pane.fv ? -1 : 1));
  xform.push(translate(-iw / 2, -ih / 2));

  const inner: Op[] = [];
  if (sc.clipActive) {
    inner.push(...composite(sc, rect));
  } else {
    inner.push(new DrawImage(sc.image, rect, "smooth"));
    if (anim.scanning) inner.push(shimmer(anim.scanPhase, rect, iw, ih));
    if (sc.segMode) {
      const s2 = seg(sc, anim, now, rect, s);
      if (anim.reveal < 1.0) {
        const z = 0.97 + 0.03 * anim.reveal;
        inner.push(
          new Push(
            [new Push(s2, scaleAbout(z, z, iw / 2, ih / 2))],
            [],
            Math.max(0.0, anim.reveal),
          ),
        );
      } else {
        inner.push(...s2);
      }
    }
  }

  ops.push(new Push(inner, xform));
  return new DisplayList(ops, s, [vw, vh], anim.needsTick());
}
